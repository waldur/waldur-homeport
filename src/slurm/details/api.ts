import { getAll } from '@waldur/core/api';

import { UserUsage } from './types';

export const getAllocationUserUsages = (params) =>
  getAll<UserUsage>('/slurm-allocation-user-usage/', { params });
