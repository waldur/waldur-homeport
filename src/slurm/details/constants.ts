import { translate } from '@waldur/i18n';

export const getChartSpec = () => [
  {
    name: translate('CPU usage'),
    field: 'cpu_usage',
  },
  {
    name: translate('GPU usage'),
    field: 'gpu_usage',
  },
  {
    name: translate('RAM usage'),
    field: 'ram_usage',
  },
];

export const palette = [
  '#003366',
  '#006699',
  '#23c6c8',
  '#eee8d5',
  '#1c84c6',
  'green',
  'red',
  'orange',
  'yellow',
  'purple',
  'aqua',
  'pink',
];
