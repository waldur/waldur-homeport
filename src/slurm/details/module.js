import slurmAllocationDetailsDialog from './slurm-allocation-details-dialog';
import slurmAllocationUsageChart from './slurm-allocation-usage-chart';
import slurmAllocationUsageTable from './slurm-allocation-usage-table';
import SlurmAllocationUsageService from './slurm-allocation-usage-service';
import { SlurmAllocationSummary } from './SlurmAllocationSummary';
import * as ResourceSummary from '@waldur/resource/summary/registry';

export default module => {
  ResourceSummary.register('SLURM.Allocation', SlurmAllocationSummary);
  module.component(
    'slurmAllocationDetailsDialog',
    slurmAllocationDetailsDialog,
  );
  module.directive('slurmAllocationUsageChart', slurmAllocationUsageChart);
  module.component('slurmAllocationUsageTable', slurmAllocationUsageTable);
  module.service('SlurmAllocationUsageService', SlurmAllocationUsageService);
};
