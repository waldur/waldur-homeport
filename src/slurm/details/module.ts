import * as ResourceSummary from '@waldur/resource/summary/registry';

import slurmAllocationDetailsDialog from './slurm-allocation-details-dialog';
import slurmAllocationUsageChart from './slurm-allocation-usage-chart';
import slurmAllocationUsageTable from './slurm-allocation-usage-table';
import { SlurmAllocationSummary } from './SlurmAllocationSummary';

export default (module) => {
  ResourceSummary.register('SLURM.Allocation', SlurmAllocationSummary);
  module.component(
    'slurmAllocationDetailsDialog',
    slurmAllocationDetailsDialog,
  );
  module.directive('slurmAllocationUsageChart', slurmAllocationUsageChart);
  module.component('slurmAllocationUsageTable', slurmAllocationUsageTable);
};
