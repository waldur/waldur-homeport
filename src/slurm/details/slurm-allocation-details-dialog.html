<div class="modal-header">
  <h3 class="modal-title" translate>SLURM allocation usage</h3>
</div>
<div>
  <div class="modal-body">
    <slurm-allocation-usage-table
      resource="$ctrl.resolve.resource"
    ></slurm-allocation-usage-table>
  </div>
  <div class="modal-footer">
    <button class="btn btn-default" type="button" ng-click="$ctrl.close()">
      <span translate>Close</span>
    </button>
  </div>
</div>
