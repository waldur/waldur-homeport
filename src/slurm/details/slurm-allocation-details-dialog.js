import template from './slurm-allocation-details-dialog.html';

const slurmAllocationDetailsDialog = {
  template,
  bindings: {
    resolve: '<',
    dismiss: '&',
    close: '&',
  },
};

export default slurmAllocationDetailsDialog;
