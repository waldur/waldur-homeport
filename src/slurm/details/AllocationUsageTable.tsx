import { FunctionComponent } from 'react';
import { PanelBody, Tab, Tabs } from 'react-bootstrap';
import { useAsync } from 'react-use';

import { EChart } from '@waldur/core/EChart';
import { LoadingSpinner } from '@waldur/core/LoadingSpinner';
import { translate } from '@waldur/i18n';

import { loadCharts } from './utils';

import './AllocationUsageTable.scss';

export const AllocationUsageTable: FunctionComponent<{ resource }> = ({
  resource,
}) => {
  const {
    loading,
    error,
    value: charts,
  } = useAsync(() =>
    loadCharts(resource.url, resource.marketplace_resource_uuid),
  );

  return loading ? (
    <LoadingSpinner />
  ) : error || !charts ? (
    <>{translate('Unable to load data')}</>
  ) : (
    <div className="tabs-container">
      <Tabs
        defaultActiveKey="tab-0"
        id="allocation-usage-tabs"
        unmountOnExit
        mountOnEnter
        animation
      >
        {charts.map((chart, index) => (
          <Tab title={chart.name} key={index} eventKey={`tab-${index}`}>
            <PanelBody>
              <EChart options={chart.options} height="350px" />
            </PanelBody>
          </Tab>
        ))}
      </Tabs>
    </div>
  );
};
