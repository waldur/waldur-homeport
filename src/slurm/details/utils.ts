import moment from 'moment';

import { translate } from '@waldur/i18n';
import { getComponentUsages } from '@waldur/marketplace/common/api';
import { getChartSpec, palette } from '@waldur/slurm/details/constants';

import { getAllocationUserUsages } from './api';
import { Period, Usage, UserUsage } from './types';

const eChartInitialOption = () => ({
  color: palette,
  tooltip: {
    trigger: 'axis',
    axisPointer: {
      type: 'cross',
      crossStyle: {
        color: '#999',
      },
    },
  },
  legend: {
    data: [translate('Usage')],
  },
  xAxis: [
    {
      type: 'category',
      data: [],
      axisPointer: {
        type: 'shadow',
      },
    },
  ],
  yAxis: [
    {
      type: 'value',
      name: translate('Usage'),
      min: 0,
      max: null,
      interval: null,
      axisLabel: {
        formatter: '{value} h',
      },
    },
    {
      type: 'value',
      name: '',
      min: 0,
      max: null,
      interval: null,
      axisLabel: {
        formatter: '{value} h',
      },
    },
  ],
  series: [
    {
      name: translate('Usage'),
      type: 'line',
      yAxisIndex: 1,
      data: [],
    },
  ],
});

const getDistinctUsers = (userUsages: UserUsage[]): UserUsage[] => {
  const distinctUsers: UserUsage[] = [];
  const map = new Map();
  for (const item of userUsages) {
    if (!map.has(item.username)) {
      map.set(item.username, true);
      distinctUsers.push(item);
    }
  }
  return distinctUsers;
};

const getLastSixMonths = (): string[] => {
  const lastSixMonths: string[] = [];
  for (let i = 5; i >= 0; i--) {
    const date = moment().subtract(i, 'months');
    const month = date.month() + 1;
    const year = date.year();
    lastSixMonths.push(`${month} - ${year}`);
  }
  return lastSixMonths;
};

export const getPeriods = (xAxisPeriods: string[]): Period[] => {
  const periods: Period[] = [];
  xAxisPeriods.forEach((period) => {
    periods.push({
      month: parseInt(period.split(' - ')[0]),
      year: parseInt(period.split(' - ')[1]),
    });
  });
  return periods;
};

const convertMBToGB = (mb: number): number =>
  parseFloat((mb / 1024).toFixed(2));

const convertMinuteToHour = (minutes: number): number =>
  parseFloat((minutes / 60).toFixed(2));

const setMaxAndAverageUsages = (option, usages: Usage[], chart) => {
  let maxUsage = Math.max(...usages.map((usage) => usage[chart.field]));
  let averageUsage =
    usages.reduce((total, next) => total + next[chart.field], 0) /
    usages.length;

  if (chart.field === 'ram_usage') {
    maxUsage = convertMBToGB(maxUsage);
    averageUsage = convertMBToGB(averageUsage);
  }

  maxUsage = convertMinuteToHour(maxUsage);
  averageUsage = convertMinuteToHour(averageUsage);

  option.yAxis[0].max = option.yAxis[1].max = Math.ceil(
    maxUsage + averageUsage / 3,
  );
  option.yAxis[0].interval = option.yAxis[1].interval = averageUsage;
};

const setUnitForRamUsage = (option, chart) => {
  if (chart.field === 'ram_usage') {
    option.yAxis[0].axisLabel.formatter = option.yAxis[1].axisLabel.formatter =
      '{value} GB-hours';
  }
};

const fillUsages = (option, periods: Period[], usages: Usage[], chart) => {
  // filling data of line chart (general usages)
  for (let i = 0; i < periods.length; i++) {
    for (let j = 0; j < usages.length; j++) {
      if (
        periods[i].month === usages[j].month &&
        periods[i].year === usages[j].year
      ) {
        let usageValue = usages[j][chart.field];
        if (chart.field === 'ram_usage') {
          usageValue = convertMBToGB(usageValue);
        }
        usageValue = convertMinuteToHour(usageValue);
        option.series[0].data.push(usageValue);
        break;
      }
      if (j === usages.length - 1) {
        option.series[0].data.push(0);
      }
    }
  }
};

const fillUserUsages = (option, periods, userUsages, chart) => {
  // filling data arrays of specific users
  for (let i = 1; i < option.series.length; i++) {
    for (let j = 0; j < periods.length; j++) {
      for (let k = 0; k < userUsages.length; k++) {
        if (
          option.series[i].name === userUsages[k].username &&
          periods[j].month === userUsages[k].month &&
          periods[j].year === userUsages[k].year
        ) {
          let usageValue = userUsages[k][chart.field];
          if (chart.field === 'ram_usage') {
            usageValue = convertMBToGB(usageValue);
          }
          usageValue = convertMinuteToHour(usageValue);
          option.series[i].data.push(usageValue);
          break;
        }
        if (k === userUsages.length - 1) {
          option.series[i].data.push(0);
        }
      }
    }
  }
};

const filterUsagesBySixMonthsPeriod = (usages: Usage[]): Usage[] =>
  usages.filter((usage) => {
    const sixMonthsAgo = moment().subtract(6, 'months');
    const usageDate = moment()
      .date(1)
      .month(usage.month - 1)
      .year(usage.year);
    return sixMonthsAgo <= usageDate && usageDate <= moment();
  });

const fillSeriesAndLegendWithDistinctUsers = (
  option,
  userUsages: UserUsage[],
) => {
  const distinctUsers = getDistinctUsers(userUsages);
  distinctUsers.forEach((user) => {
    option.series.push({
      name: user.username,
      type: 'bar',
      data: [],
      yAxisIndex: 0,
    });
    option.legend.data.push(user.username);
  });
};

export const getEChartOptions = (
  chart,
  usages: Usage[],
  userUsages: UserUsage[],
) => {
  const option = eChartInitialOption();

  // filling periods
  option.xAxis[0].data = getLastSixMonths();
  const periods = getPeriods(option.xAxis[0].data);

  usages = filterUsagesBySixMonthsPeriod(usages);
  setMaxAndAverageUsages(option, usages, chart);
  setUnitForRamUsage(option, chart);
  fillUsages(option, periods, usages, chart);
  fillSeriesAndLegendWithDistinctUsers(option, userUsages);
  fillUserUsages(option, periods, userUsages, chart);

  return option;
};

export const loadCharts = async (
  allocationUrl: string,
  resourceUuid: string,
) => {
  const componentUsages = await getComponentUsages(resourceUuid);
  const periodUsages = {};
  componentUsages.forEach((component) => {
    const period = moment(component.billing_period).format('YYYY-MM');
    periodUsages[period] = periodUsages[period] || {};
    periodUsages[period][component.type] = component.usage;
  });
  const usages: Usage[] = Object.keys(periodUsages).map((period) => {
    const date = moment(period, 'YYYY-MM');
    return {
      year: date.year(),
      month: date.month() + 1,
      cpu_usage: periodUsages[period].cpu,
      gpu_usage: periodUsages[period].gpu,
      ram_usage: periodUsages[period].ram,
    };
  });
  const userUsages = await getAllocationUserUsages({
    allocation: allocationUrl,
  });
  return getChartSpec().map((chart) => ({
    ...chart,
    options: getEChartOptions(chart, usages, userUsages),
  }));
};
