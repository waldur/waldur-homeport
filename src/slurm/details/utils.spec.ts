import { advanceTo, clear } from 'jest-date-mock';

import * as marketplaceApi from '@waldur/marketplace/common/api';
import * as slurmApi from '@waldur/slurm/details/api';

import { loadCharts } from './utils';

const componentUsages = require('./fixtures/component-usages.json');
const userUsages = require('./fixtures/user-usages.json');

jest.mock('@waldur/slurm/details/api');
jest.mock('@waldur/marketplace/common/api');

const slurmApiMock = slurmApi as jest.Mocked<typeof slurmApi>;
const marketplaceApiMock = marketplaceApi as jest.Mocked<typeof marketplaceApi>;

describe('SLURM allocation usage chart formatter', () => {
  beforeEach(() => {
    advanceTo(new Date(2020, 6, 1));
    slurmApiMock.getAllocationUserUsages.mockResolvedValue(userUsages);
    marketplaceApiMock.getComponentUsages.mockResolvedValue(componentUsages);
  });

  afterEach(() => {
    clear();
  });

  it('parses data and returns eChart option correctly', async () => {
    const charts = await loadCharts('allocationUrl', 'resourceUuid');
    expect(charts).toMatchSnapshot();
  });
});
