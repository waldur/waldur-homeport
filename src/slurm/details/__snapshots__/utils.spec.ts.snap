// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SLURM allocation usage chart formatter parses data and returns eChart option correctly 1`] = `
Array [
  Object {
    "field": "cpu_usage",
    "name": "CPU usage",
    "options": Object {
      "color": Array [
        "#003366",
        "#006699",
        "#23c6c8",
        "#eee8d5",
        "#1c84c6",
        "green",
        "red",
        "orange",
        "yellow",
        "purple",
        "aqua",
        "pink",
      ],
      "legend": Object {
        "data": Array [
          "Usage",
          "douglas",
          "steven",
          "corey",
        ],
      },
      "series": Array [
        Object {
          "data": Array [
            16.67,
            33.33,
            66.67,
            33.33,
            0,
            0,
          ],
          "name": "Usage",
          "type": "line",
          "yAxisIndex": 1,
        },
        Object {
          "data": Array [
            0,
            0,
            25,
            0,
            16.67,
            0,
          ],
          "name": "douglas",
          "type": "bar",
          "yAxisIndex": 0,
        },
        Object {
          "data": Array [
            0,
            16.67,
            0,
            66.67,
            16.67,
            0,
          ],
          "name": "steven",
          "type": "bar",
          "yAxisIndex": 0,
        },
        Object {
          "data": Array [
            0,
            0,
            8.33,
            0,
            0,
            0,
          ],
          "name": "corey",
          "type": "bar",
          "yAxisIndex": 0,
        },
      ],
      "tooltip": Object {
        "axisPointer": Object {
          "crossStyle": Object {
            "color": "#999",
          },
          "type": "cross",
        },
        "trigger": "axis",
      },
      "xAxis": Array [
        Object {
          "axisPointer": Object {
            "type": "shadow",
          },
          "data": Array [
            "2 - 2020",
            "3 - 2020",
            "4 - 2020",
            "5 - 2020",
            "6 - 2020",
            "7 - 2020",
          ],
          "type": "category",
        },
      ],
      "yAxis": Array [
        Object {
          "axisLabel": Object {
            "formatter": "{value} h",
          },
          "interval": 37.5,
          "max": 80,
          "min": 0,
          "name": "Usage",
          "type": "value",
        },
        Object {
          "axisLabel": Object {
            "formatter": "{value} h",
          },
          "interval": 37.5,
          "max": 80,
          "min": 0,
          "name": "",
          "type": "value",
        },
      ],
    },
  },
  Object {
    "field": "gpu_usage",
    "name": "GPU usage",
    "options": Object {
      "color": Array [
        "#003366",
        "#006699",
        "#23c6c8",
        "#eee8d5",
        "#1c84c6",
        "green",
        "red",
        "orange",
        "yellow",
        "purple",
        "aqua",
        "pink",
      ],
      "legend": Object {
        "data": Array [
          "Usage",
          "douglas",
          "steven",
          "corey",
        ],
      },
      "series": Array [
        Object {
          "data": Array [
            16.67,
            33.33,
            66.67,
            33.33,
            0,
            0,
          ],
          "name": "Usage",
          "type": "line",
          "yAxisIndex": 1,
        },
        Object {
          "data": Array [
            0,
            0,
            25,
            0,
            16.67,
            0,
          ],
          "name": "douglas",
          "type": "bar",
          "yAxisIndex": 0,
        },
        Object {
          "data": Array [
            0,
            16.67,
            0,
            66.67,
            16.67,
            0,
          ],
          "name": "steven",
          "type": "bar",
          "yAxisIndex": 0,
        },
        Object {
          "data": Array [
            0,
            0,
            8.33,
            0,
            0,
            0,
          ],
          "name": "corey",
          "type": "bar",
          "yAxisIndex": 0,
        },
      ],
      "tooltip": Object {
        "axisPointer": Object {
          "crossStyle": Object {
            "color": "#999",
          },
          "type": "cross",
        },
        "trigger": "axis",
      },
      "xAxis": Array [
        Object {
          "axisPointer": Object {
            "type": "shadow",
          },
          "data": Array [
            "2 - 2020",
            "3 - 2020",
            "4 - 2020",
            "5 - 2020",
            "6 - 2020",
            "7 - 2020",
          ],
          "type": "category",
        },
      ],
      "yAxis": Array [
        Object {
          "axisLabel": Object {
            "formatter": "{value} h",
          },
          "interval": 37.5,
          "max": 80,
          "min": 0,
          "name": "Usage",
          "type": "value",
        },
        Object {
          "axisLabel": Object {
            "formatter": "{value} h",
          },
          "interval": 37.5,
          "max": 80,
          "min": 0,
          "name": "",
          "type": "value",
        },
      ],
    },
  },
  Object {
    "field": "ram_usage",
    "name": "RAM usage",
    "options": Object {
      "color": Array [
        "#003366",
        "#006699",
        "#23c6c8",
        "#eee8d5",
        "#1c84c6",
        "green",
        "red",
        "orange",
        "yellow",
        "purple",
        "aqua",
        "pink",
      ],
      "legend": Object {
        "data": Array [
          "Usage",
          "douglas",
          "steven",
          "corey",
        ],
      },
      "series": Array [
        Object {
          "data": Array [
            0.02,
            0.03,
            0.07,
            0.03,
            0,
            0,
          ],
          "name": "Usage",
          "type": "line",
          "yAxisIndex": 1,
        },
        Object {
          "data": Array [
            0,
            0,
            0.02,
            0,
            0.02,
            0,
          ],
          "name": "douglas",
          "type": "bar",
          "yAxisIndex": 0,
        },
        Object {
          "data": Array [
            0,
            0.02,
            0,
            0.07,
            0.02,
            0,
          ],
          "name": "steven",
          "type": "bar",
          "yAxisIndex": 0,
        },
        Object {
          "data": Array [
            0,
            0,
            0.01,
            0,
            0,
            0,
          ],
          "name": "corey",
          "type": "bar",
          "yAxisIndex": 0,
        },
      ],
      "tooltip": Object {
        "axisPointer": Object {
          "crossStyle": Object {
            "color": "#999",
          },
          "type": "cross",
        },
        "trigger": "axis",
      },
      "xAxis": Array [
        Object {
          "axisPointer": Object {
            "type": "shadow",
          },
          "data": Array [
            "2 - 2020",
            "3 - 2020",
            "4 - 2020",
            "5 - 2020",
            "6 - 2020",
            "7 - 2020",
          ],
          "type": "category",
        },
      ],
      "yAxis": Array [
        Object {
          "axisLabel": Object {
            "formatter": "{value} GB-hours",
          },
          "interval": 0.04,
          "max": 1,
          "min": 0,
          "name": "Usage",
          "type": "value",
        },
        Object {
          "axisLabel": Object {
            "formatter": "{value} GB-hours",
          },
          "interval": 0.04,
          "max": 1,
          "min": 0,
          "name": "",
          "type": "value",
        },
      ],
    },
  },
]
`;
